python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -896,6 +896,116 @@ def corr(self, method="pearson", min_periods=1):
 
         return res
 
+    def cov(self, min_periods=0, ddof=1):
+        """
+        Compute pairwise covariance of columns, excluding NA/null values.
+
+        Parameters
+        ----------
+        min_periods: int, optional
+            Minimum number of observations required per pair of columns
+            to have a valid result.
+
+        ddof: int, optional
+            Delta degrees of freedom, default is 1.
+
+        Returns
+        ----------
+        DataFrame
+            Covariance matrix.
+
+        Examples
+        --------
+        >>> import cudf
+        >>> gdf = cudf.DataFrame({
+        ...             "id": ["a", "a", "a", "b", "b", "b", "c", "c", "c"],
+        ...             "val1": [5, 4, 6, 4, 8, 7, 4, 5, 2],
+        ...             "val2": [4, 5, 6, 1, 2, 9, 8, 5, 1],
+        ...             "val3": [4, 5, 6, 1, 2, 9, 8, 5, 1]})
+        >>> gdf
+        id  val1  val2  val3
+        0  a     5     4     4
+        1  a     4     5     5
+        2  a     6     6     6
+        3  b     4     1     1
+        4  b     8     2     2
+        5  b     7     9     9
+        6  c     4     8     8
+        7  c     5     5     5
+        8  c     2     1     1
+        >>> gdf.groupby("id").cov()
+                    val1       val2       val3
+        id
+        a  val1  1.000000   0.500000   0.500000
+           val2  0.500000   1.000000   1.000000
+           val3  0.500000   1.000000   1.000000
+        b  val1  4.333333   3.500000   3.500000
+           val2  3.500000  19.000000  19.000000
+           val3  3.500000  19.000000  19.000000
+        c  val1  2.333333   3.833333   3.833333
+           val2  3.833333  12.333333  12.333333
+           val3  3.833333  12.333333  12.333333
+        """
+
+        # create expanded dataframe consisting all combinations of the
+        # struct columns-pairs to be correlated
+        # i.e (('col1', 'col1'), ('col1', 'col2'), ('col2', 'col2'))
+        column_names = self.grouping.values.columns.tolist()
+        len_cols = len(column_names)
+
+        column_pair_structs = {}
+        for x, y in itertools.combinations_with_replacement(column_names, 2):
+            column_pair_structs[(x, y)] = cudf.DataFrame._from_data(
+                {"x": self.obj._data[x], "y": self.obj._data[y]}
+            ).to_struct()
+        column_pair_groupby = cudf.DataFrame._from_data(
+            column_pair_structs
+        ).groupby(by=self.grouping.keys)
+
+        try:
+            gb_cov = column_pair_groupby.agg(
+                lambda x: x.cov(min_periods, ddof)
+            )
+        except RuntimeError as e:
+            if "Unsupported groupby reduction type-agg combination" in str(e):
+                raise TypeError(
+                    "Covariance accepts only numerical column-pairs"
```

Comment:
Should we move this error to C++ code?
(But it will be a RuntimeError. not TypeError).

Reply:
I'm not sure I understand your suggestion. This is already catching a C++ `std::runtime_error` (which is propagated as a `RuntimeError` to Python) and rethrowing it as a `TypeError`. How is what you are proposing different?

Reply:
Ignore the types for now.
"Unsupported groupby reduction type-agg combination" is coming from reduction aggregation due to dependency on mean, and sum.
This code changes the error message. If the intention of this code is to provide "Unsupported groupby reduction type-agg combination" error message instead of "Unsupported groupby reduction type-agg combination", we can move it into libcudf code to make it provide meaningful error message.

Reply:
See also #10200. Cross-referencing because this is a great example of how we could improve libcudf error handling.